core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,423 @@
+	{ 384, 0x98 },	 { 416, 0x95 },	  { 448, 0x99 },  { 480, 0x96 },
+	{ 512, 0x9A },	 { 576, 0x9B },	  { 640, 0xA0 },  { 704, 0x9D },
+	{ 768, 0xA1 },	 { 832, 0x9E },	  { 896, 0xA2 },  { 960, 0x67 },
+	{ 1024, 0xA3 },	 { 1152, 0xA4 },  { 1280, 0xA8 }, { 1536, 0xA9 },
+	{ 1792, 0xAA },	 { 1920, 0xA7 },  { 2048, 0xAB }, { 2304, 0xAC },
+	{ 2560, 0xB0 },	 { 3072, 0xB1 },  { 3584, 0xB2 }, { 3840, 0xAF },
+	{ 4096, 0xB3 },	 { 4608, 0xB4 },  { 5120, 0xB8 }, { 6144, 0xB9 },
+	{ 7168, 0xBA },	 { 7680, 0xB7 },  { 8192, 0xBB }, { 9216, 0xBC },
+	{ 10240, 0xBD }, { 12288, 0xBE }, { 15360, 0xBF }
+};
+
+/*
+ * I2C divisor and ibfd register values when glitch filter is disabled.
+ * In case of duplicate SCL divisor value, the ibfd value with high MUL value
+ * has been selected. A higher MUL value results in a lower sampling rate of
+ * the I2C signals. This gives the I2C module greater immunity against glitches
+ * in the I2C signals.
+ */
+static const struct i2c_clock_divisor_pair
+	i2c_clock_divisor_glitch_disabled[] = {
+	{ 20, 0x0 },	 { 22, 0x1 },	  { 24, 0x2 },	  { 26, 0x3 },
+	{ 28, 0x8 },	 { 30, 0x5 },	  { 32, 0x9 },	  { 34, 0x6 },
+	{ 36, 0x0A },	 { 40, 0x40 },	  { 44, 0x41 },	  { 48, 0x42 },
+	{ 52, 0x43 },	 { 56, 0x48 },	  { 60, 0x45 },	  { 64, 0x49 },
+	{ 68, 0x46 },	 { 72, 0x4A },	  { 80, 0x80 },	  { 88, 0x81 },
+	{ 96, 0x82 },	 { 104, 0x83 },	  { 112, 0x88 },  { 120, 0x85 },
+	{ 128, 0x89 },	 { 136, 0x86 },	  { 144, 0x8A },  { 160, 0x8B },
+	{ 176, 0x8C },	 { 192, 0x90 },	  { 208, 0x56 },  { 224, 0x91 },
+	{ 240, 0x1F },	 { 256, 0x92 },	  { 272, 0x8F },  { 288, 0x93 },
+	{ 320, 0x98 },	 { 352, 0x95 },	  { 384, 0x99 },  { 416, 0x96 },
+	{ 448, 0x9A },	 { 480, 0x5F },	  { 512, 0x9B },  { 576, 0x9C },
+	{ 640, 0xA0 },	 { 768, 0xA1 },	  { 896, 0xA2 },  { 960, 0x9F },
+	{ 1024, 0xA3 },	 { 1152, 0xA4 },  { 1280, 0xA8 }, { 1536, 0xA9 },
+	{ 1792, 0xAA },	 { 1920, 0xA7 },  { 2048, 0xAB }, { 2304, 0xAC },
+	{ 2560, 0xAD },	 { 3072, 0xB1 },  { 3584, 0xB2 }, { 3840, 0xAF },
+	{ 4096, 0xB3 },	 { 4608, 0xB4 },  { 5120, 0xB8 }, { 6144, 0xB9 },
+	{ 7168, 0xBA },	 { 7680, 0xB7 },  { 8192, 0xBB }, { 9216, 0xBC },
+	{ 10240, 0xBD }, { 12288, 0xBE }, { 15360, 0xBF }
+};
+
+/*
+ * ERR009203 :   I2C may not work reliably with the default setting
+ * Description : The clocking circuitry of I2C module may not work reliably due
+ * to the slow rise time of SCL signal.
+ * Workaround :  Enable the receiver digital filter by setting IBDBG[GLFLT_EN]
+ * to 1.
+ */
+static void i2c_erratum_a009203(vaddr_t base)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+
+	i2c_io_or8((vaddr_t)&regs->ibdbg, I2C_IBDBG_GLFLT_EN);
+}
+
+/*
+ * software reset of the entire I2C module.
+ * The module is reset and disabled.
+ * Status register fields (IBSR) are cleared.
+
+ * @param[in] Base       Base Address of I2c controller's registers
+ */
+void i2c_reset(vaddr_t base)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
```

Comment:
Could `io_mask8()` be used instead of `i2c_io_or8()` and `i2c_io_and8()`?
If not perhaps we should add the missing functions to `<io.h>`.

Reply:
yes
`i2c_io_or8()` can be replaced with `io_setbits8()`
`i2c_io_and8()` can be replaced with `io_clrbits8()`